src/main/java/lux/fm/bookingservice/service/impl/BookingServiceImpl.java

```diff
@@ -29,9 +31,56 @@ public class BookingServiceImpl implements BookingService {
     private final BookingRepository bookingRepository;
     private final BookingMapper bookingMapper;
     private final AccommodationRepository accommodationRepository;
+    private final UserRepository userRepository;
     private final NotificationService notificationService;
     private final PaymentRepository paymentRepository;
 
+    @Override
+    @Scheduled(cron = "0 0 0 * * ?")
+    @Transactional
+    public void checkExpiredBookings() {
+        List<Booking> expiredBookings = bookingRepository.checkExpiredBookings(
+                LocalDate.now().plusDays(1)
+        );
+        List<User> telegramUsers = userRepository.findAllByTelegramIdIsNotNull();
+
+        if (telegramUsers.isEmpty()) {
+            return;
+        }
+
+        if (expiredBookings.isEmpty()) {
+            String message = "No expired bookings today!";
+            telegramUsers.stream()
+                    .map(User::getTelegramId)
+                    .forEach(chatId -> notificationService.notifyUser(chatId, message));
+            return;
+        }
+
+        telegramUsers.stream()
+                .map(User::getTelegramId)
+                .forEach(chatId -> {
+                    expiredBookings.stream()
+                            .forEach(booking -> {
+                                booking.setStatus(Status.EXPIRED);
+                                Accommodation accommodation = booking.getAccommodation();
```

Comment:
You do it inside another stream, So for each user will be a iterating through all expired bookings and you do a lot of work. Take it out separately